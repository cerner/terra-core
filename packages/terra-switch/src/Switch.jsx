import React, { useRef, useCallback } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import { FormattedMessage } from 'react-intl';
import { KEY_SPACE, KEY_RETURN } from 'keycode-js';
import { removeFocusStyles, restoreFocusStyles } from './_SwitchUtils';
import styles from './Switch.module.scss';

const cx = classNamesBind.bind(styles);

const SWITCH_STATE = Object.freeze({
  ON: <FormattedMessage id="Terra.switch.switchstatuslabel.on" />,
  OFF: <FormattedMessage id="Terra.switch.switchstatuslabel.off" />,
});

const propTypes = {
  /**
   * Whether or not the Switch is checked ("ON").
   */
  isChecked: PropTypes.bool,
  /**
   * Whether or not the Switch is disabled.
   */
  isDisabled: PropTypes.bool,
  /**
   * The label text of the Switch component.
   */
  labelText: PropTypes.string.isRequired,
  /**
   * Callback function when switch value changes from ON / OFF.
   * Returns Parameters: 1. switch value 2. event.
   */
  onChange: PropTypes.func,
};

const defaultProps = {
  isChecked: false,
  isDisabled: false,
  onChange: undefined,
};

const Switch = (props) => {
  const {
    isChecked,
    isDisabled,
    onChange,
    labelText,
    ...customProps
  } = props;
  const theme = React.useContext(ThemeContext);
  const sliderButton = useRef();

  const handleOnClick = useCallback((event) => {
    // Need this for focus styles on IE running on OS X since it does not receive focus when clicked.
    sliderButton.current.focus();
    if (onChange) {
      onChange(!isChecked, event);
    }
  }, [isChecked, onChange]);

  const handleOnKeyDown = (event) => {
    if (event.nativeEvent.keyCode === KEY_RETURN || event.nativeEvent.keyCode === KEY_SPACE) {
      event.preventDefault();
      restoreFocusStyles(sliderButton.current);
      if (onChange) {
        onChange(!isChecked, event);
      }
    }
  };

  const handleOnMouseDown = (event) => {
    event.preventDefault();
    removeFocusStyles(sliderButton.current);
  };

  const handleOnBlur = () => {
    restoreFocusStyles(sliderButton.current);
  };

  const statusLabelText = isChecked ? SWITCH_STATE.ON : SWITCH_STATE.OFF;

  const switchClassNames = classNames(cx(
    'switch',
    { 'is-enabled': !isDisabled },
    { 'is-disabled': isDisabled },
    { 'is-selected': isChecked },
    theme.className,
  ),
  customProps.className);

  let switchAttrs;
  if (!isDisabled) {
    switchAttrs = {
      tabIndex: '0',
      onBlur: handleOnBlur,
      onClick: handleOnClick,
      onMouseDown: handleOnMouseDown,
      onKeyDown: handleOnKeyDown,
    };
  } else {
    switchAttrs = {
      'aria-disabled': true,
    };
  }

  return (
    <div
      {...customProps}
      {...switchAttrs}
      aria-label={labelText}
      aria-checked={isChecked}
      role="switch"
      className={switchClassNames}
      data-terra-switch-show-focus-styles
      ref={sliderButton}
    >
      <div aria-hidden className={cx('label-container')}>
        <div className={cx('label-text')}>{labelText}</div>
        <div className={cx('status-label-text')}>{statusLabelText}</div>
      </div>
      <div aria-hidden className={cx('tray')}>
        <div className={cx('slider')} />
      </div>
    </div>
  );
};

Switch.propTypes = propTypes;
Switch.defaultProps = defaultProps;

export default Switch;
