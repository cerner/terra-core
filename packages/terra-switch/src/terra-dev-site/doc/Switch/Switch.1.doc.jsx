import React from 'react';
import DocTemplate from 'terra-doc-template';
import ReadMe from '../../../../docs/README.md';
import { name } from '../../../../package.json';

// Component Source
import SwitchButtonSrc from '!raw-loader!../../../../src/SwitchButton';
import SwitchCheckBoxSrc from '!raw-loader!../../../../src/SwitchCheckBox';

// Example Files
import DefaultSwitchButton from '../example/DefaultSwitchButton';
import DefaultSwitchButtonSrc from '!raw-loader!../../../../src/terra-dev-site/doc/example/DefaultSwitchButton.jsx';
import DefaultSwitchCheckBox from '../example/DefaultSwitchCheckBox';
import DefaultSwitchCheckBoxSrc from '!raw-loader!../../../../src/terra-dev-site/doc/example/DefaultSwitchCheckBox.jsx';

const DocPage = () => (
  <DocTemplate
    packageName={name}
    readme={ReadMe}
    srcPath={`https://github.com/cerner/terra-core/tree/master/packages/${name}`}
    examples={[
      {
        title: 'Switch With Button',
        example: <DefaultSwitchButton />,
        source: DefaultSwitchButtonSrc,
      },
      {
        title: 'Switch With Checkbox',
        example: <DefaultSwitchCheckBox />,
        source: DefaultSwitchCheckBoxSrc,
      },
    ]}
    propsTables={[
      {
        componentName: 'Switch-Button',
        componentSrc: SwitchButtonSrc,
      },

      {
        componentName: 'Switch-Checkbox',
        componentSrc: SwitchCheckBoxSrc,
      },
    ]}
  />
);

export default DocPage;
