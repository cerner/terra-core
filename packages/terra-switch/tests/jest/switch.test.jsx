import React from 'react';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { shallowWithIntl } from 'terra-enzyme-intl';
import Switch from '../../src/Switch';

describe('Switch', () => {
  const defaultRender = <Switch labelText="default" />;

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallowWithIntl(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as disabled when set', () => {
    const button = shallowWithIntl(<Switch labelText="Label" isDisabled />);
    expect(button).toMatchSnapshot();
    expect(button.prop('aria-disabled')).toBe(true);
  });

  // Prop Tests
  it('should use the default label text when is given', () => {
    const wrapper = shallowWithIntl(defaultRender);
    expect(wrapper.find('.label-text').text()).toEqual('default');
  });

  // Structure Tests
  it('should have the class switch', () => {
    const wrapper = shallowWithIntl(defaultRender);
    expect(wrapper.prop('className')).toContain('switch');
  });
});
