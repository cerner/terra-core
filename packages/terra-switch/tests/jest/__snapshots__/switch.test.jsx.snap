// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render a default component 1`] = `
<div
  aria-checked={false}
  aria-label="default"
  className="switch is-enabled"
  data-terra-switch-show-focus-styles={true}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="switch"
  tabIndex="0"
>
  <div
    aria-hidden={true}
    className="label-container"
  >
    <div
      className="label-text"
    >
      default
    </div>
    <div
      className="status-label-text"
    >
      <FormattedMessage
        id="Terra.switch.switchstatuslabel.off"
        values={Object {}}
      />
    </div>
  </div>
  <div
    aria-hidden={true}
    className="tray"
  >
    <div
      className="slider"
    />
  </div>
</div>
`;

exports[`Switch should render as disabled when set 1`] = `
<div
  aria-checked={false}
  aria-disabled={true}
  aria-label="Label"
  className="switch is-disabled"
  data-terra-switch-show-focus-styles={true}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  role="switch"
>
  <div
    aria-hidden={true}
    className="label-container"
  >
    <div
      className="label-text"
    >
      Label
    </div>
    <div
      className="status-label-text"
    >
      <FormattedMessage
        id="Terra.switch.switchstatuslabel.off"
        values={Object {}}
      />
    </div>
  </div>
  <div
    aria-hidden={true}
    className="tray"
  >
    <div
      className="slider"
    />
  </div>
</div>
`;
